packages/starlight/integrations/expressive-code.ts

```diff
@@ -0,0 +1,265 @@
+				if (!configuredThemesSupportDarkModeSwitch) {
+					if (useStarlightDarkModeSwitch === true)
+						console.warn(
+							`*** Warning: To support the config option "useStarlightDarkModeSwitch: true", please provide 1 dark and 1 light theme (found ${foundDarkThemes} and ${foundLightThemes}).\n`
+						);
+					return '';
+				}
+				return darkModeSwitchCss.join('\n');
+			},
+		});
+	}
+
+	if (useStarlightUiThemeColors === true && configuredThemesCount < 2)
+		console.warn(`*** Warning: Using the config option "useStarlightUiThemeColors: true" with only one theme is not recommended. For better color contrast, please provide 1 dark and 1 light theme.\n`);
+
+	// Add Expressive Code UI translations (if any) for all defined locales
+	if (locales) {
+		for (const locale in locales) {
+			const lang = locales[locale]?.lang;
+			if (!lang) continue;
+
+			addTranslations(lang, pluginFramesTexts, [
+				'expressiveCode.copyButtonCopied',
+				'expressiveCode.copyButtonTooltip',
+				'expressiveCode.terminalWindowFallbackTitle',
+			]);
+		}
+	}
+
+	return [
+		astroExpressiveCode({
+			theme,
+			customizeTheme: (theme) => {
+				if (useStarlightUiThemeColors) {
+					applyStarlightUiThemeColors(theme);
+				}
+				if (customizeTheme) {
+					theme = customizeTheme(theme) ?? theme;
+				}
+				if (minSyntaxHighlightingColorContrast > 0) {
+					ensureMinTokenColorContrast(theme, minSyntaxHighlightingColorContrast);
+				}
+				return theme;
+			},
+			useThemedSelectionColors,
+			styleOverrides: {
+				codeFontFamily: 'var(--__sl-font-mono)',
+				codeFontSize: 'var(--sl-text-code)',
+				codeLineHeight: 'var(--sl-line-height)',
+				uiFontFamily: 'var(--__sl-font)',
+				textMarkers: {
+					defaultChroma: '45',
+					backgroundOpacity: '60%',
+					...textMarkersStyleOverrides,
+				},
+				...otherStyleOverrides,
+			},
+			getBlockLocale: ({ file }) => {
+				// Root path:    `src/content/docs/getting-started.mdx`
+				// Locale path:  `src/content/docs/fr/getting-started.mdx`
+				// Part indices:  0     1      2   3         4
+				const pathParts = path.relative(file.cwd, file.path).split(/[\\/]/);
+				const localeOrFolder = pathParts[3];
+				const lang = localeOrFolder ? locales?.[localeOrFolder]?.lang : locales?.root?.lang;
+				const defaultLang = defaultLocale?.lang || defaultLocale?.locale;
+				const result = lang || defaultLang || 'en';
+
+				return result;
+			},
+			plugins,
+			...rest,
+		}),
+	];
+};
+
+function addTranslations<T extends PluginTexts<any>>(
+	lang: string,
+	target: T,
+	ids: (keyof (typeof translations)['en'])[]
+) {
+	const builtinTexts = translations[lang];
+	let userTexts: Partial<(typeof translations)['en']> = {};
+	try {
+		// As Astro has not been initialized yet, we cannot use content collections directly
+		// and have to manually load any user-provided translations from the file system
+		userTexts = JSON.parse(fs.readFileSync(`src/content/i18n/${lang}.json`, 'utf8'));
```

Comment:
Refactored the translation system in #1003, which is now merged, so I added a commit to this PR which addresses this comment — EC now uses the same translations system we already use internally in Starlight.

Reply:
That's awesome, thank you so much! My architectural changes to the new efficient multi-theme rendering based on CSS variables are nearing completion as well. I just got all my existing tests green today and now only need to add a few new tests and perform finalization & cleanup work. Then the biggest change to EC since its first release on NPM is done and I can finally get this PR ready to merge! Looking really forward to that. Thanks everyone for your patience!